### ISSUE TITLE
id_token signature validation

### ISSUE THREAD

**Issue by Peperud** [grafana/grafana#15312]
Grafana does not seem to have a configuration setting for the signing public key of the token server. Nor does it have a setting to specify an OpenID compliant discovery URL. 
This would indicate that it does not properly validate identity tokens as it would need the above to validate the signature.

**Comment by DanCech** [comment]
That is correct, validation could be implemented in `pkg/social/generic_oauth.go` using the `gopkg.in/square/go-jose.v2` library which is already vendored.

**Comment by daniellee** [comment]
@DanCech how critical is this? Is this a potential security issue?

**Comment by DanCech** [comment]
It isn't, because you're already validating the endpoint by connecting to it via SSL.  Validating the token signature is just an additional layer.  It isn't a bad idea but there is no difference in terms of security between accepting the data encoded in the id token vs using the id token to request the user info (which doesn't contain any additional signature).

The OpenID implementer's guide speaks about validating the content of the id token, but doesn't mention validation of the signature itself. https://openid.net/specs/openid-connect-basic-1_0.html#IDTokenValidation

In any case, this code is actually supporting a non-standard extension of OAuth 2.0 used by Microsoft Azure AD which returns userinfo in the token, this isn't trying to validate the token as an OpenID token.

Further discussion in the original PRs #10560 and #11127

At this point my recommendation is that we should add a strict OpenID Connect implementation as a separate module, so that users with authentication servers that properly support OpenID Connect can use that while maintaining the more flexible implementation in the generic oauth module for the many non-compliant servers out there.

**Comment by Peperud** [comment]
Strictly speaking OAuth2 by itself does not define an authentication protocol at all. That's why a lot of companies did their own thing on top of it to use it for authentication. And that's where OpenID stepped in to provide the missing authentication piece and standardize it. 
Grafana's generic oauth module is using **id_token** (an OpenID only artifact) and is somewhat blurring the lines.

I agree - a dedicated  OpenID Connect compliant implementation should be added.

**Comment by mwxp06** [comment]
Any progress?

**Comment by Jguer** [comment]
ID Token signature validation has been implemented in the Azure OIDC provider

**Comment by nicolas17** [comment]
Maybe many things requested in #12312, #27443 and #26350 can be coalesced into "Add a proper OpenID Connect authentication method instead of having bits and pieces of OpenID Connect bolted onto the generic OAuth2 method".
